### PR TITLE
Add various resource constraints for StringMax/StringMin

### PR DESCRIPTION
**Pull request by ashemedai**
*Description of changes:*

Based on the exchange in #2797 with @kddejong I whipped up a small script to scan the various resources for `Name` and `Description` and see if they were missing their constraints. This is the result of that.

For `Description` if an explicit maximum was given, but no minimum I reasoned that `0` is valid since it's only descriptive and an empty string is okay. For `Name`, as it is typically (always) an identifier it should always have a minimum of `1`. Is this a fair assumption?

I initially wanted to add the `AllowedPatternRegexp` information, but the syntax at first seems all over the place in the CloudFormation documentation.

There are still some mistakes in this that I need to rectify, but they also depend on my question above.

I also have not done any checking against botocore to reference their specified constraints for this. What do we prefer, reference their data to avoid repeating and potential discrepencies?

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by kddejong** [comment]
@ashemedai  Interesting script any chance you would be willing to share it.  There are min/max numbers, regex patterns, string min/max, allowed values would be interesting to see what more I could pull.

We do test regex patterns to validate they are valid with python regex package. We now use regex package instead of re as it provides more support for some of the regex patterns that we are using.

**Comment by ashemedai** [comment]
@kddejong That was the intention at least. :)

Let me toss it up as a gist tomorrow.

**Comment by kddejong** [comment]
I made a fix do you want to look at this pr #2824

File: src/cfnlint/data/ExtendedSpecs/all/03_value_types/aws_cassandra.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
+[
+ {
+  "op": "add",
+  "path": "/ValueTypes/AWS::Cassandra::Keyspace.KeyspaceName",
+  "value": {
+   "StringMax": 255,
```
Comment (kddejong) [inline src/cfnlint/data/ExtendedSpecs/all/03_value_types/aws_cassandra.json:6]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cassandra-keyspace.html#cfn-cassandra-keyspace-keyspacename


We are pulling this regex which doesn't match the described length.

File: src/cfnlint/data/ExtendedSpecs/all/03_value_types/aws_cassandra.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+   "StringMin": 3
+  }
+ },
+ {
+  "op": "add",
+  "path": "/ValueTypes/AWS::Cassandra::Table.TableName",
```
Comment (kddejong) [inline src/cfnlint/data/ExtendedSpecs/all/03_value_types/aws_cassandra.json:12]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cassandra-table.html#cfn-cassandra-table-tablename

Same we have the pattern that defines the length and the pattern and length descriptions don't match.

File: src/cfnlint/data/ExtendedSpecs/all/03_value_types/aws_athena.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
+[
+ {
+  "op": "add",
+  "path": "/ValueTypes/AWS::Athena::CapacityReservation.Name",
+  "value": {
+   "StringMax": 128,
```
Comment (kddejong) [inline src/cfnlint/data/ExtendedSpecs/all/03_value_types/aws_athena.json:6]: Covered by regex:
```
"AWS::Athena::CapacityReservation.Name": {
   "AllowedPatternRegex": "[a-zA-Z0-9._-]{1,128}"
  },
```

File: src/cfnlint/data/ExtendedSpecs/all/03_value_types/aws_athena.json [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+   "StringMin": 1
+  }
+ },
+ {
+  "op": "add",
+  "path": "/ValueTypes/AWS::Athena::WorkGroup.Name",
+  "value": {
+   "StringMax": 128,
```
Comment (kddejong) [inline src/cfnlint/data/ExtendedSpecs/all/03_value_types/aws_athena.json:14]: Covered by regex:

```
"AWS::Athena::WorkGroup.Name": {
   "AllowedPatternRegex": "[a-zA-Z0-9._-]{1,128}"
  },
```

File: src/cfnlint/data/ExtendedSpecs/all/03_value_types/aws_appflow.json [hunk 5]
```diff
@@ -0,0 +1,82 @@
+[
+ {
+  "op": "add",
+  "path": "/ValueTypes/AWS::AppFlow::Connector.Description",
+  "value": {
+   "StringMax": 1024,
```
Comment (kddejong) [inline src/cfnlint/data/ExtendedSpecs/all/03_value_types/aws_appflow.json:6]: Need to check this one.  The schema defines 2048 and the docs have 1024

"Description": {
   "description": "A description about the connector that's being registered.",
   "maxLength": 2048,
   "pattern": "[\\s\\w/!@#+=.-]*",
   "type": "string"
  }

**Comment by ashemedai** [comment]
@kddejong I put the current version at https://gist.github.com/ashemedai/962d3449c3712e40bb54d2fa3fba4e00

Things I would like to do:

- Properly update the project json files, I think my current code should allow it to run within the cfn-lint project, but I need to adjust the output path properly, maybe a parameter?
- Depending on how many parameters need to be added, use `argparse`.
- Tighten up existing regular expressions.
- Add regex matches for some other common cases.
- Maybe parse out/check for the `{x,y}` bits at the end of some regexp so we know to skip `StringMax`/`StringMin` since the regexp already takes care of it.

File: src/cfnlint/data/ExtendedSpecs/all/03_value_types/aws_athena.json (hunk 3 above)
Comment (ashemedai) [reply src/cfnlint/data/ExtendedSpecs/all/03_value_types/aws_athena.json:6]: So when the regexp covers it, it overrides the need for any explicit `StringMax`/`StringMin`? In these case I checked the documentation and filled these in based on my reading of the regexp.

File: src/cfnlint/data/ExtendedSpecs/all/03_value_types/aws_redshift.json [hunk 6]
```diff
@@ -6,5 +6,45 @@
...
+  "value": {
+   "StringMax": 2147483647,
+   "StringMin": 0
+  }
+ },
+ {
+  "op": "add",
+  "path": "/ValueTypes/AWS::Redshift::EventSubscription.SubscriptionName",
+  "value": {
+   "StringMax": 2147483647,
```
Comment (blade2005) [inline src/cfnlint/data/ExtendedSpecs/all/03_value_types/aws_redshift.json:30]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-redshift-eventsubscription.html#cfn-redshift-eventsubscription-subscriptionname
While the printed text says the max is 2147483647 the conditions says the max is 255. I think additional fact finding may be needed on some of these.

I've found inconsistencies with different services in how the constraints are describe and quite often conflict with each other or omit a constraint and the api throws an error.

I would bet whoever wrote the docs for cloudformation just put the 32 bit signed int max and wrote in their docs more refined constraints. Like `AWS::Redshift::EventSubscription.SourceType` says it's max length is `2147483647` but it's a valid values are: cluster, cluster-parameter-group, cluster-security-group, cluster-snapshot, and scheduled-action.

**Comment by kddejong** [comment]
Lots of conflicts now as we have converted to v1.  Lets see how and where we can add these changes as needed.

**Pull request closed without merging**